Sources/Datasets/datasets/structured/Iris.swift

```diff
@@ -0,0 +1,147 @@
+    public static let numberOfTrainingExamples = 150
+    public static let numberOfFeatures = 4
+
+    private init() {}
+}
+
+fileprivate let irisInfo = S5TFDatasetInfo(
+    name: "Iris",
+    version: "0.0.1",
+    description: """
+    This is perhaps the best known database to be found in the pattern recognition
+    literature. Fisher's paper is a classic in the field and is referenced
+    frequently to this day. (See Duda & Hart, for example.) The data set contains
+    3 classes of 50 instances each, where each class refers to a type of iris
+    plant. One class is linearly separable from the other 2; the latter are NOT
+    linearly separable from each other.
+    """,
+    homepage: URL(string: "http://archive.ics.uci.edu/ml")!
+)
+
+public struct IrisDataLoader: S5TFDataLoader {
+    private var index = 0
+
+    let batchSize: Int
+    let data: Tensor<Float>
+    let labels: Tensor<Int32>
+
+    // MARK: - Initializers.
+    private init(batchSize: Int) {
+        self.batchSize = batchSize
+
+        // Download necessary files.
+        let semaphore = DispatchSemaphore(value: 0)
+        var csvUrl: URL? = nil
+        print("Fetching files... Waiting for the download to finish before continuing...")
+        Downloader().download(fileAt: URL(string: "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data")!,
+            cacheName: "iris", fileName: "iris.csv") { url, err in
+            guard let url = url else {
+                fatalError("URL not saved.")
+            }
+            csvUrl = url
+            semaphore.signal()
+        }
+        semaphore.wait()
+
+        // Load data.
+        let rawData = try! String(contentsOfFile: csvUrl!.absoluteString)
+        // Use a flattened array because Swift does not support 2 dimensional arrays for initialization.
+        var XValues = [Float]()
+        var yValues = [Int32]()
+        for line in rawData.split(separator: "\n") {
+        let items = line.split(separator: ",")
+            // Load features.
+            // TOOD: use broadcasting when it's available.
+            for j in 0...3 {
+                XValues.append(Float(items[j])!)
+            }
+
+            // Load labels.
+            switch items[4] {
+            case "Iris-setosa": yValues.append(0)
+            case "Iris-versicolor": yValues.append(1)
+            case "Iris-virginica": yValues.append(2)
+            default: fatalError("Unkown label: \(items[4])")
+            }
+        }
+
+        // Convert so Swift Tensors and store on self.
+        self.data = Tensor<Float>(XValues).reshaped(to: TensorShape(Iris.numberOfTrainingExamples, Iris.numberOfFeatures))
+        self.labels = Tensor<Int32>(yValues)
+    }
+
+    private init(batchSize: Int, data: Tensor<Float>, labels: Tensor<Int32>) {
+        self.batchSize = batchSize
+        self.data = data
+        self.labels = labels
+    }
+
+    public init() {
+        self.init(batchSize: 1)
```

Comment:
I think we should init with `batchSize: 0`.

Reply:
That would cause an infinite loop when looping over it without `.batched`. Shall we add a check to `next` (`DataLoader`)?

Reply:
Yeah, that sounds good.

Reply:
Will be solved in https://github.com/s5tf-team/datasets/issues/17.

Reply:
See https://github.com/s5tf-team/S5TF/pull/13.